src/torchjd/autojac/_utils.py

```diff
@@ -30,3 +30,31 @@ def _check_retain_graph_compatible_with_chunk_size(
             "When using `retain_graph=False`, parameter `parallel_chunk_size` must be `None` or "
             "large enough to compute all gradients in parallel."
         )
+
+
+def _get_leafs_of_autograd_graph(roots: list[Tensor], excluded: set[Tensor]) -> set[Tensor]:
+    """
+    Gets the leafs of the autograd graph of all tensors in `roots`, excludes leafs that are in the
+    autograd graph of any tensor in `excluded`.
+    :param roots: Roots of the autograd graphs.
+    :param excluded: Tensors excluded from the search.
+    :return: A list of tensors representing the variables that lead to the given node.
+    """
+    node_stack = [tensor.grad_fn for tensor in roots if tensor not in excluded]
+    excluded_nodes = {tensor.grad_fn for tensor in excluded}
```

Comment:
The grad_fn of leafs is None, we can exclude it from this list, alternatively, we could partition if they are leafs.

Reply:
I don't understand, could you reformulate?